File: Beetle.NetPackage/Sources/c#/Controller.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Beetle.NetPackage
{
    public class Controller
    {
        private static Dictionary<Type, ControllerAction> mActions = new Dictionary<Type, ControllerAction>();

        public static void Invoke(Object actionobj, NetClient client, Object message)
        {
            lock (mActions)
            {

                Type key = actionobj.GetType();
                ControllerAction action = null;
                if (!mActions.TryGetValue(key, out action))
                {
                    action = new ControllerAction(key);
                    mActions[key] = action;
                }
                action.Invoke(actionobj, client, message);
            }
        }
    }
}

File: Beetle.NetPackage/Sources/android/Controller.java
package beetle.netpackage;

import java.util.HashMap;

public class Controller {

	private static HashMap<Class<?>, ControllerAction> mActions = new HashMap<Class<?>, ControllerAction>();

	public static void Invoke(Object actionobj, NetClient client, Object message) {
		synchronized (mActions) {

			Class<?> key = actionobj.getClass();
			ControllerAction action = mActions.get(key);
			if (action == null) {
				action = new ControllerAction(key);
				mActions.put(key, action);
			}
			action.Invoke(actionobj, client, message);
		}
	}
}
